src/librustc_incremental/persist/load.rs

```diff
@@ -277,21 +262,37 @@ fn dirty_nodes<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
                    hash.dep_node);
         }
 
-        dirty_nodes.insert(hash.dep_node.clone());
+        dirty_nodes.insert(hash.dep_node.clone(), hash.dep_node.clone());
     }
 
     dirty_nodes
 }
 
+fn transitive_dirty_nodes(edges: &[SerializedEdge],
+                          mut dirty_nodes: DirtyNodes)
+                          -> DirtyNodes
+{
+    let mut len = 0;
+    while len != dirty_nodes.len() {
+        len = dirty_nodes.len();
+        for edge in edges {
```

Comment:
Yeah, I was lazy. :) I can rewrite it to use a work-list or some such thing. One thing is that we don't have the edges indexed by their target, which we would want here. Perhaps I'll change how things are serialized to be a `Map<Target, Vec<Source>>`. I think we even build one of those in the "reduction" algorithm, so perhaps I should just return that (i.e., don't return a `Graph`, but some kind of `ReducedGraph`). I have to look at how the code works there.